btree/btree.c

```diff
@@ -2952,6 +2961,518 @@ static void ff_node_free_credits(const struct nd *node,
+	 *
+	 *        keysize + valsize <= vkv_space(slot->s_node)
+	 */
+}
+
+static void vkv_fid(const struct nd *node, struct m0_fid *fid)
+{
+	
+}
+
+static void *vkv_key(const struct nd *node, int idx)
+{
+	/**
+	 * @brief Return the memory address pointing to key space.
+	 * 
+	 * index = given index
+	 * start_addr = node->n_addr + sizeof(node_header)
+	 * new_key = start_addr + dir[index].key_offset
+	 * return new_key
+	 */
+}
+
+static void *vkv_val(const struct nd *node, int idx)
+{
+	/**
+	 * @brief Return the memory address pointing to value space.
+	 * 
+	 * index = given index
+	 * start_addr =  node->n_addr + segaddr_shift(node->n_addr) - 1
+	 * new_val = start_addr - dir[index].val_offset
+	 * return new_val
+	 */
+}
+
+static void vkv_rec(struct slot *slot)
+{
+	/**
+	 * @brief This function will fill the provided slot with its key and
+	 *        value based on the index provided.
+	 *        vkv_key() and vkv_val() functions should be used to get the
+	 *        memory address pointing to key and val.
+	 *        vkv_keysize() and vkv_valsize() functions should be used to
+	 *        fill the v_count parameter of the slot vector.
+	 */
+}
+
+static void vkv_node_key(struct slot *slot)
+{
+	/**
+	 * @brief This function will fill the provided slot with its key and
+	 *        value based on the index provided.
+	 *        vkv_key() functions should be used to get the
+	 *        memory address pointing to key and val.
+	 *        vkv_keysize() functions should be used to
+	 *        fill the v_count parameter of the slot vector.
+	 */
+}
+
+static void vkv_child(struct slot *slot, struct segaddr *addr)
+{
+	/**
+	 * @brief This function will return the memory address pointing to its
+	 *        child node. We will call the function vkv_val() for this
+	 *        purpose.
+	 *        This function is called for internal nodes.
+	 */
+}
+
+static bool vkv_isfit(struct slot *slot)
+{
+	/**
+	 * @brief This function will determine if the current key and value
+	 *        can be added to the node.
+	 *        It also checks the validity of the record.
+	 *        Modifications will be required in nt_isfit() function
+	 *        declaration as we need to have the sizes of incoming key and
+	 *        value.
+	 *
+	 *        keysize + valsize <= vkv_space(slot->s_node)
+	 */
+}
+
+static void vkv_done(struct slot *slot, struct m0_be_tx *tx, bool modified)
+{
+
+}
+
+static void vkv_make(struct slot *slot, struct m0_be_tx *tx)
+{
+	/**
+	 * @brief This function will create space to add new entry at the
+	 *        given index. It is assumed that the possibility whether this
+	 *        can occur or not is determined before calling this function.
+	 *
+	 *        start_key_addr = vkv_key(slot->s_node, index)
+	 *        start_val_addr = vkv_val(slot->s_node, index)
+	 *        count = num of records i.e. vkv_used
+	 *
+	 *        idx = index
+	 *        while(idx <= count) {
+	 *          total_key_size += vkv_keysize(slot->s_node, idx)
```

Comment:
to find out total_key_size, you can directly find out offset of last key and size of it , so lets addr = offset of last key + size  and the difference of offset key at idx and addr will be the total_key_size , No need to use while loop. Similarly logic can be applied for total_val_size.

Reply:
Updated.